### PR TITLE
Move tests to the hack repo

### PR DESCRIPTION
**Pull request by chizhg**
Also move the test files to the hack repo.

/cc @n3wscott

### CODE REVIEW

File: test/unit/presubmit-tests.sh [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    (( ! POST_UNIT_TESTS )) || test_failed "Post unit tests did run"
+    (( ! PRE_INTEGRATION_TESTS )) || test_failed "Pre integration tests did run"
+    (( ! RAN_INTEGRATION_TESTS )) || test_failed "Custom integration tests did run"
+    (( ! POST_INTEGRATION_TESTS )) || test_failed "Post integration tests did run"
+    echo "Test failed fast"
+  }
+}
+
+function test_exempt_md() {
+  source $(dirname $0)/../../scripts/presubmit-tests.sh
```
Comment (n3wscott) [inline test/unit/presubmit-tests.sh:190]: This is no longer in the scripts dir.

Search the rest of the code for scripts?

File: test/unit/presubmit-tests.sh [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+PROW_JOB_ID=123
+JOB_TYPE="presubmit"
+ARTIFACTS=/tmp
+
+source $(dirname $0)/test-helper.sh
+
+set -e
+
+function init_test_env() {
+  source $(dirname $0)/../../scripts/presubmit-tests.sh
```
Comment (n3wscott) [inline test/unit/presubmit-tests.sh:29]: And here

**Review by n3wscott — COMMENTED** [review]
Some missed paths on scripts

File: test/unit/presubmit-tests.sh (hunk 1 above)
Comment (chizhg) [reply test/unit/presubmit-tests.sh]: Done.

File: test/unit/presubmit-tests.sh (hunk 2 above)
Comment (chizhg) [reply test/unit/presubmit-tests.sh]: Done.

**Comment by chizhg** [comment]
Please note I had to remove `//+build hack` to make `go test ./...` not panic.

File: hack.go [hunk 3]
```diff
@@ -1,5 +1,3 @@
-// +build hack
```
Comment (n3wscott) [inline hack.go:1]: We really do want to have this tag here.

File: test/e2e/simple_e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+package e2e
+
+import (
+	"testing"
+)
+
+// This is actually a unit test, but it does exercise the necessary helper functions.
+
+func TestE2ESucceeds(t *testing.T) {
```
Comment (n3wscott) [inline test/e2e/simple_e2e_test.go:27]: What is the point of these? We should drop them from this repo. They should not be in the hack repo.

File: test/library_test.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import (
+	"fmt"
+	"os"
+	"syscall"
+	"testing"
+	"time"
+)
+
+// signal sends a UNIX signal to the test process.
+func signal(s os.Signal) {
```
Comment (n3wscott) [inline test/library_test.go:30]: What is the point of these? We should drop them from this repo. They should not be in the hack repo.

Comment (chizhg) [reply test/library_test.go:30]: This is for unit testing the `library.sh` script.

File: test/e2e/simple_e2e_test.go (hunk 4 above)
Comment (chizhg) [reply test/e2e/simple_e2e_test.go:27]: This is for the smoke e2e tests to make sure the base e2e test flow is not broken.

File: hack.go (hunk 3 above)
Comment (chizhg) [reply hack.go:1]: What's the reason behind it? We had https://github.com/knative/test-infra/blob/master/scripts/dummy.go which doesn't have build tags.

Comment (n3wscott) [reply hack.go:1]: It was a mistake to not have build tags on the original file. We are using this file for go mod control, but no code should be here.

Comment (chizhg) [reply hack.go]: Got it, I have added it back.

**Review by n3wscott — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**